File: backend/scripts/verify_mapping_suggest.py
"""
Walkthrough: auto-suggest is frontend-only; upload + GET must not persist.

Mirrors frontend mappingSuggestions.js (case-insensitive exact match).
Bootstraps the same in-memory SQLite isolation as tests/conftest.py
(does NOT touch docgen.db / live Postgres).

Run from docgen/backend:
  python scripts/verify_mapping_suggest.py
"""
from __future__ import annotations

import io
import os
import shutil
import sys
import tempfile
from pathlib import Path

# Isolate before any app imports — same contract as tests/conftest.py
os.environ["DATABASE_URL"] = "sqlite:///:memory:"
os.environ["JWT_SECRET"] = (
    "test_jwt_secret_for_platform_isolation_tests_only_64chars_xx"
)
os.environ["ENVIRONMENT"] = "development"
os.environ["ALLOW_DEMO_SEED"] = "false"
os.environ["DOCGEN_SKIP_PDF"] = "true"
os.environ["PLATFORM_SIGNUP_RATE_LIMIT"] = "1000/hour"
os.environ["PLATFORM_LOGIN_RATE_LIMIT"] = "1000/minute"
os.environ["PLATFORM_INVITE_RATE_LIMIT"] = "1000/hour"

_TEST_FILE_ROOT = Path(tempfile.mkdtemp(prefix="docgen_suggest_walk_"))
_TEST_TEMPLATE_DIR = _TEST_FILE_ROOT / "template_store"
_TEST_OUTPUT_DIR = _TEST_FILE_ROOT / "output"
_TEST_TEMPLATE_DIR.mkdir(parents=True, exist_ok=True)
_TEST_OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
os.environ["TEMPLATE_DIR"] = str(_TEST_TEMPLATE_DIR)
os.environ["OUTPUT_DIR"] = str(_TEST_OUTPUT_DIR)

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

from docx import Document  # noqa: E402
from fastapi.testclient import TestClient  # noqa: E402
from sqlalchemy import create_engine, event  # noqa: E402
from sqlalchemy.orm import sessionmaker  # noqa: E402
from sqlalchemy.pool import StaticPool  # noqa: E402

import database  # noqa: E402
from database import Base, get_db  # noqa: E402

_engine = create_engine(
    "sqlite://",
    connect_args={"check_same_thread": False},
    poolclass=StaticPool,
)


@event.listens_for(_engine, "connect")
def _fk_on(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute("PRAGMA foreign_keys=ON")
    cursor.close()


database.engine = _engine
database.SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=_engine)

import models  # noqa: E402,F401
from auth import create_org_jwt, hash_password  # noqa: E402
from main import app  # noqa: E402
from models import Organization, OrgUser, User  # noqa: E402

SessionLocal = database.SessionLocal
Base.metadata.create_all(bind=_engine)


def suggest_field_key(placeholder_key: str, resolvable_keys: list[str]) -> str | None:
    needle = (placeholder_key or "").lower()
    if not needle:
        return None
    for key in resolvable_keys:
        if str(key).lower() == needle:
            return key
    return None


def build_initial_selections(detected, persisted, resolvable_keys):
    saved = {
        row["placeholder_key"]: row["field_key"]
        for row in persisted
        if row.get("is_mapped") and row.get("placeholder_key") and row.get("field_key")
    }
    selections = {}
    suggested = []
    for key in detected:
        if key in saved:
            selections[key] = saved[key]
            continue
        match = suggest_field_key(key, resolvable_keys)
        if match:
            selections[key] = match
            suggested.append(key)
        else:
            selections[key] = ""
    return selections, suggested


def _docx_bytes(*placeholders: str) -> bytes:
    doc = Document()
    doc.add_paragraph(" ".join(f"{{{{{p}}}}}" for p in placeholders))
    buf = io.BytesIO()
    doc.save(buf)
    return buf.getvalue()


def _make_org_admin(db, *, slug: str, username: str, org_name: str):
    user = User(
        username=username,
        full_name=f"{org_name} Admin",
        password_hash=hash_password("test-password-123"),
        role="staff",
        is_active=True,
    )
    db.add(user)
    db.flush()
    org = Organization(name=org_name, slug=slug, is_active=True)
    db.add(org)
    db.flush()
    membership = OrgUser(org_id=org.id, user_id=user.id, role="org_admin")
    db.add(membership)
    db.commit()
    token = create_org_jwt(user_id=user.id, org_id=org.id, role="org_admin")
    return {"token": token, "org": org}


def main() -> int:
    db = SessionLocal()
    try:
        org = _make_org_admin(
            db,
            slug="suggest-walk",
            username="suggest_admin",
            org_name="Suggest Walk Org",
        )
    finally:
        db.close()

    def override_get_db():
        session = SessionLocal()
        try:
            yield session
        finally:
            session.close()

    app.dependency_overrides[get_db] = override_get_db
    client = TestClient(app)
    client.cookies.set("platform_access_token", org["token"])

    try:
        # Pure logic
        assert suggest_field_key("Company_Name", ["company_name"]) == "company_name"
        assert suggest_field_key("Other_Thing", ["company_name"]) is None

        dt = client.post(
            "/api/platform/document-types/",
            json={"name": "Suggest Walk", "slug": "suggest-walk"},
        )
        assert dt.status_code == 201, dt.text
        dt_id = dt.json()["id"]

        flow = client.post(f"/api/platform/{dt_id}/flow", json={})
        assert flow.status_code == 201, flow.text
        flow_id = flow.json()["id"]

        step = client.post(
            f"/api/platform/{flow_id}/steps",
            json={
                "step_type": "text_field",
                "order_index": 0,
                "label": "Company",
                "is_enabled": True,
            },
        )
        assert step.status_code == 201, step.text
        step_id = step.json()["id"]

        field = client.post(
            f"/api/platform/steps/{step_id}/fields",
            json={
                "field_key": "company_name",
                "field_label": "Company name",
                "field_type": "text",
                "is_required": True,
            },
        )
        assert field.status_code == 201, field.text

        pub = client.post(f"/api/platform/{flow_id}/publish")
        assert pub.status_code == 200, pub.text

        upload = client.post(
            f"/api/platform/{dt_id}/templates",
            files={
                "file": (
                    "suggest.docx",
                    _docx_bytes("company_name", "Other_Thing"),
                    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
                )
            },
        )
        assert upload.status_code == 201, upload.text
        template_id = upload.json()["id"]

        before = client.get(f"/api/platform/templates/{template_id}/mappings")
        assert before.status_code == 200, before.text
        body = before.json()
        detected = body["detected_placeholders"]
        assert "company_name" in detected
        assert "Other_Thing" in detected
        mapped = [m for m in body.get("mappings") or [] if m.get("is_mapped")]
        assert mapped == [], "GET after upload must not persist mappings"

        selections, suggested = build_initial_selections(
            detected, body.get("mappings") or [], ["company_name"]
        )
        assert selections["company_name"] == "company_name"
        assert "company_name" in suggested
        assert selections["Other_Thing"] == ""
        assert "Other_Thing" not in suggested

        after = client.get(f"/api/platform/templates/{template_id}/mappings")
        assert [
            m for m in after.json().get("mappings") or [] if m.get("is_mapped")
        ] == [], "still unsaved after UI-suggest simulation"

        save = client.post(
            f"/api/platform/templates/{template_id}/mappings",
            json={
                "mappings": [
                    {"placeholder_key": "company_name", "field_key": "company_name"},
                ]
            },
        )
        assert save.status_code == 200, save.text
        assert any(
            m.get("is_mapped") and m["placeholder_key"] == "company_name"
            for m in save.json().get("mappings") or []
        )

        print(
            {
                "matching": "case-insensitive exact",
                "suggested_visual": "blue AntD Tag 'Suggested'",
                "preselected": "company_name",
                "empty": "Other_Thing",
                "auto_post_on_load": False,
                "explicit_save": True,
                "ok": True,
            }
        )
        return 0
    finally:
        app.dependency_overrides.clear()
        shutil.rmtree(_TEST_FILE_ROOT, ignore_errors=True)


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend/scripts/verify_mapping_suggest.mjs
/**
 * Scripted walkthrough: auto-suggest mapping (frontend pure logic) +
 * API proof that GET mappings does not persist (no auto-POST).
 *
 * Run from frontend/: node scripts/verify_mapping_suggest.mjs
 * Optional API leg: set PLATFORM_API_BASE (default http://127.0.0.1:8000/api/platform)
 * and ensure backend is up; otherwise pure-logic checks still run.
 */
import assert from 'node:assert/strict'
import { readFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath, pathToFileURL } from 'node:url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const suggestionsUrl = pathToFileURL(
  join(__dirname, '../src/pages/platform/mappingSuggestions.js')
).href

const {
  suggestFieldKeyForPlaceholder,
  buildInitialMappingSelections,
} = await import(suggestionsUrl)

function section(title) {
  console.log(`\n=== ${title} ===`)
}

section('1. Matching logic (case-insensitive exact)')
assert.equal(
  suggestFieldKeyForPlaceholder('company_name', ['company_name', 'date']),
  'company_name'
)
assert.equal(
  suggestFieldKeyForPlaceholder('Company_Name', ['company_name', 'date']),
  'company_name',
  'case-insensitive: Company_Name → company_name'
)
assert.equal(
  suggestFieldKeyForPlaceholder('DATE', ['company_name', 'date']),
  'date'
)
assert.equal(
  suggestFieldKeyForPlaceholder('Other_Thing', ['company_name', 'date']),
  null,
  'no match → null'
)
assert.equal(
  suggestFieldKeyForPlaceholder('company', ['company_name']),
  null,
  'no substring / fuzzy match'
)
assert.equal(
  suggestFieldKeyForPlaceholder('country.name', ['country.name', 'party.email']),
  'country.name',
  'fixed step outputs match'
)
console.log('OK suggestFieldKeyForPlaceholder')

section('2. Initial selections — suggest vs empty vs persisted')
const { selections, suggestedKeys } = buildInitialMappingSelections({
  detected: ['company_name', 'Other_Thing', 'Date', 'already_saved'],
  persistedMappings: [
    { placeholder_key: 'already_saved', field_key: 'date', is_mapped: true },
  ],
  resolvableKeys: ['company_name', 'date', 'country.name'],
})
assert.equal(selections.company_name, 'company_name')
assert.ok(suggestedKeys.includes('company_name'))
assert.equal(selections.Other_Thing, '')
assert.ok(!suggestedKeys.includes('Other_Thing'))
assert.equal(selections.Date, 'date')
assert.ok(suggestedKeys.includes('Date'))
assert.equal(selections.already_saved, 'date')
assert.ok(
  !suggestedKeys.includes('already_saved'),
  'persisted mapping is not tagged Suggested'
)
console.log('OK buildInitialMappingSelections', { selections, suggestedKeys })

section('3. No auto-POST on load (source contract)')
const panelSrc = readFileSync(
  join(__dirname, '../src/pages/platform/PlaceholderMappingPanel.jsx'),
  'utf8'
)
assert.ok(
  panelSrc.includes('buildInitialMappingSelections'),
  'panel uses suggest helper on load'
)
assert.ok(panelSrc.includes('Suggested'), 'Suggested tag present')
assert.ok(
  /savePlaceholderMappings\(template\.id/.test(panelSrc),
  'save POST only via save()'
)
// savePlaceholderMappings must not appear inside load/useCallback load body
// except we call listPlaceholderMappings (GET). Ensure POST helper only in save.
const loadMatch = panelSrc.match(
  /const load = useCallback\(async \(\) => \{[\s\S]*?\}, \[documentTypeId, template\.id\]\)/
)
assert.ok(loadMatch, 'found load()')
assert.ok(
  !loadMatch[0].includes('savePlaceholderMappings'),
  'load() must not call savePlaceholderMappings (no auto-POST)'
)
console.log('OK panel load does not POST mappings')

section('4. Optional live API leg (template upload + GET only)')
const apiBase = (
  process.env.PLATFORM_API_BASE || 'http://127.0.0.1:8000/api/platform'
).replace(/\/$/, '')

let apiRan = false
try {
  const health = await fetch(`${apiBase.replace(/\/platform$/, '')}/health`)
  if (!health.ok) throw new Error(`health ${health.status}`)

  const stamp = Date.now()
  const signup = await fetch(`${apiBase}/signup`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      org_name: `Suggest Org ${stamp}`,
      org_slug: `suggest-org-${stamp}`,
      admin_email: `suggest-${stamp}@example.com`,
      admin_password: 'SuggestTest1!',
      admin_full_name: 'Suggest Admin',
    }),
  })
  if (!signup.ok) {
    // try login path if signup shape differs — skip API leg
    throw new Error(`signup ${signup.status}: ${await signup.text()}`)
  }
  const signupBody = await signup.json()
  const token =
    signupBody.access_token ||
    signupBody.token ||
    signupBody.platform_access_token
  if (!token) throw new Error('no token in signup response')

  const headers = {
    Authorization: `Bearer ${token}`,
    'Content-Type': 'application/json',
  }

  const dt = await fetch(`${apiBase}/document-types/`, {
    method: 'POST',
    headers,
    body: JSON.stringify({
      name: `Suggest DT ${stamp}`,
      slug: `suggest-dt-${stamp}`,
    }),
  })
  assert.equal(dt.status, 201, await dt.text())
  const dtId = (await dt.json()).id

  const flow = await fetch(`${apiBase}/${dtId}/flow`, {
    method: 'POST',
    headers,
    body: JSON.stringify({}),
  })
  assert.equal(flow.status, 201, await flow.text())
  const flowId = (await flow.json()).id

  const step = await fetch(`${apiBase}/${flowId}/steps`, {
    method: 'POST',
    headers,
    body: JSON.stringify({
      step_type: 'text_field',
      order_index: 0,
      label: 'Company',
      is_enabled: true,
    }),
  })
  assert.equal(step.status, 201, await step.text())
  const stepId = (await step.json()).id

  const field = await fetch(`${apiBase}/steps/${stepId}/fields`, {
    method: 'POST',
    headers,
    body: JSON.stringify({
      field_key: 'company_name',
      field_label: 'Company name',
      field_type: 'text',
      is_required: true,
    }),
  })
  assert.equal(field.status, 201, await field.text())

  const pub = await fetch(`${apiBase}/${flowId}/publish`, {
    method: 'POST',
    headers,
  })
  assert.equal(pub.status, 200, await pub.text())

  // Minimal docx via python-docx is preferred; try dynamic require of a tiny zip.
  // Use backend TestClient path if fetch multipart is awkward — build minimal OOXML.
  const { Document, Packer, Paragraph, Text } = await import('docx').catch(() => ({}))
  let docxBuffer
  if (Document && Packer) {
    const doc = new Document({
      sections: [
        {
          children: [
            new Paragraph({
              children: [
                new Text('Hello {{company_name}} and {{Other_Thing}}'),
              ],
            }),
          ],
        },
      ],
    })
    docxBuffer = Buffer.from(await Packer.toBuffer(doc))
  } else {
    // Fallback: call Python one-liner via child_process
    const { spawnSync } = await import('node:child_process')
    const py = `
from io import BytesIO
from docx import Document
d = Document()
d.add_paragraph("Hello {{company_name}} and {{Other_Thing}}")
b = BytesIO(); d.save(b); import sys; sys.stdout.buffer.write(b.getvalue())
`
    const r = spawnSync('python', ['-c', py], { encoding: 'buffer', maxBuffer: 5e6 })
    if (r.status !== 0) throw new Error(r.stderr?.toString() || 'docx gen failed')
    docxBuffer = r.stdout
  }

  const form = new FormData()
  form.append(
    'file',
    new Blob([docxBuffer], {
      type: 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    }),
    'suggest_test.docx'
  )
  const upload = await fetch(`${apiBase}/${dtId}/templates`, {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}` },
    body: form,
  })
  assert.equal(upload.status, 201, await upload.text())
  const template = await upload.json()

  const before = await fetch(`${apiBase}/templates/${template.id}/mappings`, {
    headers: { Authorization: `Bearer ${token}` },
  })
  assert.equal(before.status, 200, await before.text())
  const beforeBody = await before.json()
  const detected = beforeBody.detected_placeholders || []
  assert.ok(detected.includes('company_name'))
  assert.ok(detected.includes('Other_Thing'))
  const mappedBefore = (beforeBody.mappings || []).filter((m) => m.is_mapped)
  assert.equal(
    mappedBefore.length,
    0,
    'GET after upload must not have persisted mappings (no auto-POST)'
  )

  const ui = buildInitialMappingSelections({
    detected,
    persistedMappings: beforeBody.mappings || [],
    resolvableKeys: ['company_name'],
  })
  assert.equal(ui.selections.company_name, 'company_name')
  assert.ok(ui.suggestedKeys.includes('company_name'))
  assert.equal(ui.selections.Other_Thing, '')
  assert.ok(!ui.suggestedKeys.includes('Other_Thing'))

  const after = await fetch(`${apiBase}/templates/${template.id}/mappings`, {
    headers: { Authorization: `Bearer ${token}` },
  })
  const afterBody = await after.json()
  assert.equal(
    (afterBody.mappings || []).filter((m) => m.is_mapped).length,
    0,
    'still unsaved after simulating UI suggest'
  )

  apiRan = true
  console.log('OK live API: upload + GET; suggest pre-fill; no silent persist')
} catch (err) {
  console.log(`SKIP live API leg: ${err.message}`)
}

section('DONE')
console.log(
  JSON.stringify(
    {
      matching: 'case-insensitive exact',
      suggested_visual: 'blue AntD Tag labeled Suggested',
      api_leg: apiRan ? 'ran' : 'skipped',
    },
    null,
    2
  )
)
